x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -379,20 +378,22 @@ private void updateMapping(LongConsumer handler, AtomicInteger retryCounter) {
 
     private void handleFailure(Exception e, AtomicInteger retryCounter, Runnable task) {
         assert e != null;
-        if (shouldRetry(e)) {
-            if (isStopped() == false && retryCounter.incrementAndGet() <= RETRY_LIMIT) {
-                LOGGER.debug(new ParameterizedMessage("{} error during follow shard task, retrying...", params.getFollowShardId()), e);
-                scheduler.accept(retryTimeout, task);
-            } else {
-                markAsFailed(new ElasticsearchException("retrying failed [" + retryCounter.get() +
-                    "] times, aborting...", e));
-            }
+        if (shouldRetry(e) && isStopped() == false) {
+            LOGGER.debug(new ParameterizedMessage("{} error during follow shard task, retrying...", params.getFollowShardId()), e);
+            int currentRetry = retryCounter.incrementAndGet();
+            long delay = computeDelay(currentRetry, maxRetryDelay.getMillis());
+            scheduler.accept(TimeValue.timeValueMillis(delay), task);
         } else {
             markAsFailed(e);
         }
     }
 
-    private boolean shouldRetry(Exception e) {
+    static long computeDelay(int currentRetry, long maxRetryDelayInMillis) {
+        long expectedBackOff = Math.round(10 * Math.exp(0.8d * currentRetry) - 1);
```

Comment:
This is not quite what I think of as exponential backoff. The problem that I see with an implementation like this is that we can have a thundering herd problem. If there is some failure that causes a bunch of tasks to simultaneously fail (e.g., say that we have a bunch of outstanding fetch tasks waiting for a response, and the network connection breaks, failing all of them), all of the retries will keep waking up at the same time, itself potentially causing issues due to a herd.

Typically it would be that there is a random component in exponential backoff, to avoid this herding. As a first approach, what I suggest here is: choose a random value k between 0 and 2^number of retries - 1. Then retry after k * delay seconds. We can cap this at max retry delay.

Reply:
Right, I did think about this, but when looking at `ExponentialBackoffIterator` class to see how it was implemented elsewhere I decided to use that. I will change this.

Reply:
I've changed how the delay is now computed. With this randomness the retying mechanism is much more solid.